### PR TITLE
feat: Add an option for adding a separate display name for repositories

### LINKED ISSUE #8953: Add new env var `SEMGREP_REPO_DISPLAY_NAME` to allow overriding repo name in Semgrep Cloud Platform
**Is your feature request related to a problem? Please describe.**
The value associated with `SEMGREP_REPO_NAME` is used to generate the API URL for hitting external SCM APIs to publish review comments. However, sometimes users want to override the repo name with a display name just for cosmetic purposes in the app, but this isn't currently possibly because of how `SEMGREP_REPO_NAME` is used as both the display name in the app and also used to construct the URL to hit the external API of SCMs. 

**Describe the solution you'd like**
Add a new environment variable `SEMGREP_REPO_DISPLAY_NAME` where users can specify the display name used for that particular repo on Semgrep. Fall back on `SEMGREP_REPO_NAME` if the user does not specify anything. 

**Describe alternatives you've considered**
We considered altering the behavior of how the repo name is used in the app but decided against it due to parsing reliability of repo URLs. More details in [this ticket](https://linear.app/semgrep/issue/FIND-1132/setting-semgrep-repo-url-does-not-properly-override-the-semgrep-repo)

**Use case**
What will this feature enable for you?
Allows users to group findings together under "subproject" overrides and still allow users capability to fire PR comments. 

**Additional context**

### PR DESCRIPTION
**Pull request by emjin**
Closes #8953 

People running Semgrep on monorepos may want to scan the sub repositories separately. Add a repo display name separate from the actual repo name to allow people to configure this.

Test plan: 

Added a print statement on `ci.py` line 294 to output `project_meta`. Here's the result:
```
➜  engine git:(emma/gh-8953-allow-overriding-display-name) ✗  SEMGREP_REPO_DISPLAY_NAME="EMMA_TESTING_NAME" semgrep ci --dry-run

                  
                  
┌────────────────┐
│ Debugging Info │
└────────────────┘
                  
  SCAN ENVIRONMENT
  versions    - semgrep 1.59.1 on python 3.11.7                        
  environment - running in environment git, triggering event is unknown
            
  CONNECTION
ProjectMetadata(semgrep_version=Version(value='1.59.1'), scan_environment='git', repository='semgrep', repo_url=Uri(value='https://github.com/semgrep/semgrep'), branch='emma/gh-8953-allow-overriding-display-name', commit=Sha1(value='67b81f5f1b42a594d17b87ace32b08b80186256e'), commit_title="Not sure why pre-commit didn't run", commit_author_email='emma@semgrep.com', commit_author_name='Emma Jin', commit_author_username=None, commit_author_image_url=None, ci_job_url=None, on='unknown', pull_request_author_username=None, pull_request_author_image_url=None, pull_request_id=None, pull_request_title=None, is_full_scan=True, repo_id=None, org_id=None, repo_display_name='EMMA_TESTING_NAME', commit_timestamp=Datetime(value='2024-02-12T16:23:40'), base_sha=None, start_sha=None, is_sca_scan=False, is_code_scan=False, is_secrets_scan=False)
  Initializing scan   .    
```

I also ran with just `SEMGREP_REPO_NAME` overridden to check that this also affects the display name.

```
➜  engine git:(emma/gh-8953-allow-overriding-display-name) ✗  SEMGREP_REPO_NAME="EMMA_TESTING_NAME" semgrep ci --dry-run 

                  
                  
┌────────────────┐
│ Debugging Info │
└────────────────┘
                  
  SCAN ENVIRONMENT
  versions    - semgrep 1.59.1 on python 3.11.7                        
  environment - running in environment git, triggering event is unknown
            
  CONNECTION
ProjectMetadata(semgrep_version=Version(value='1.59.1'), scan_environment='git', repository='EMMA_TESTING_NAME', repo_url=Uri(value='https://github.com/semgrep/semgrep'), branch='emma/gh-8953-allow-overriding-display-name', commit=Sha1(value='67b81f5f1b42a594d17b87ace32b08b80186256e'), commit_title="Not sure why pre-commit didn't run", commit_author_email='emma@semgrep.com', commit_author_name='Emma Jin', commit_author_username=None, commit_author_image_url=None, ci_job_url=None, on='unknown', pull_request_author_username=None, pull_request_author_image_url=None, pull_request_id=None, pull_request_title=None, is_full_scan=True, repo_id=None, org_id=None, repo_display_name='EMMA_TESTING_NAME', commit_timestamp=Datetime(value='2024-02-12T16:23:40'), base_sha=None, start_sha=None, is_sca_scan=False, is_code_scan=False, is_secrets_scan=False)
  Initializing scan  ..              
```

### CODE REVIEW

**Comment by chmccreery** [comment]
Right idea, but this particular function is only used by `semgrep scan` and not `semgrep ci` and we will need this in both. I'm looking right now for the best place to handle this.

Edit: Ok, it looks like to_project_metadata inside of the `GitMeta` class also needs this information. So you could either include a property in the GitMeta (following the same pattern as repo_url) that fetches its value from the environment variable, or falls back on the `SEMGREP_REPO_NAME` if the display name is not set. Let me know if you want to chat in person 👍

**Comment by emjin** [comment]
Not sure about changelog---it is user facing, but it doesn't do anything until we put in the app changes. Leaning towards no

Edit: Decided to do one after all

File: cli/src/semgrep/meta.py [hunk 1]
```diff
@@ -113,6 +113,11 @@ def repo_name(self) -> str:
         repo_root_str = rev_parse.stdout.strip()
         return str(os.path.basename(repo_root_str))
 
+    @property
+    def repo_display_name(self) -> str:
+        repo_display_name = os.getenv("SEMGREP_REPO_DISPLAY_NAME", self.repo_name)
+        return repo_display_name or self.repo_name
```
Comment (chmccreery) [inline cli/src/semgrep/meta.py]: Redundant to have fallback on both of these lines, but doesn't hurt 😄

File: cli/src/semgrep/meta.py [hunk 2]
```diff
@@ -659,6 +670,11 @@ def repo_name(self) -> str:
             return project_path
         return super().repo_name
 
+    @property
+    def repo_display_name(self) -> str:
+        repo_display_name = os.getenv("SEMGREP_REPO_DISPLAY_NAME", self.repo_name)
+        return repo_display_name or self.repo_name
```
Comment (chmccreery) [inline cli/src/semgrep/meta.py]: probably cleaner to have all of these (besides the one in the GitMeta class) call the super().repo_display_name method, which will keep this logic centralized in one place. Then the rest of these classes will inherit that behavior from GitMeta 👍 

```suggestion
    @property
    def repo_display_name(self) -> str:
        return super().repo_display_name
```

**Review by chmccreery — COMMENTED** [review]
> I also ran with just SEMGREP_REPO_NAME overridden to check that this also affects the display name.

Is the goal to make this non-nullable in the interface some day? We could also handle this on the app side, and I'm trying to decide if there's any argument for doing so (the only thing I can think of is that we would know whether or not the user explicitly set this environment variable - and I'm not sure that's important).

**Comment by emjin** [comment]
We could make it non-nullable, we'd just have to wait 10 versions first because of our backwards compatibility promises. I believe that for semgrep ci runs both semgrep_repo_name and semgrep_repo_display_name must end up with a value.

File: cli/src/semgrep/meta.py (hunk 1 above)
Comment (emjin) [reply cli/src/semgrep/meta.py]: lol I was resolving pre-commit and didn't think 😂

File: cli/src/semgrep/meta.py [hunk 3]
```diff
@@ -203,6 +208,7 @@ def to_project_metadata(self) -> out.ProjectMetadata:
             # REQUIRED for semgrep-app backend
             repository=self.repo_name,
             # OPTIONAL for semgrep-app backend
+            repo_display_name=self.repo_display_name,
```
Comment (chmccreery) [inline cli/src/semgrep/meta.py:210]: nit: I think this should move up a line and be considered "required" if we are going to have it fall back on self.repo_name. Alternatively, you could have it not do that, and be of type `str | None` and leave it here.

**Review by chmccreery — APPROVED** [review]

**Pull request merged**